src/proto/payload.proto

```diff
@@ -1,8 +1,12 @@
-syntax = "proto3";
-package pb;
+syntax = "proto2";
```

Comment:
please leave syntax to be proto3, protons library is not supporting proto2 officially and can sometimes generate weird results. In proto3 fields are optional by default so you can omit optional keywords.

Reply:
This is on purpose. The spec recently changed to go to proto2 since proto3 has some unwanted semantics: https://github.com/libp2p/specs/pull/456

Reply:
I understand that but it also says it's compatible with proto3. And I would rather not switch to proto2 because of and conclusion there: https://github.com/ipfs/protons/issues/34

Reply:
> `optional NoiseExtensions extensions = 4;` is represented as a oneof in proto3

go-libp2p's older noise `.proto` file [with protobuf v3](https://github.com/libp2p/go-libp2p/blob/a82be49036096c93f44287780a8d32142b591976/p2p/security/noise/pb/payload.proto) doesn't have the `oneof` keyword.  I don't think it's implicit if not set so I'm not sure this is correct.  Am I missing something?

> but it's a bit of a hack.

Is this because the field identifier is non-sequential?  I think this is a feature of protocol buffers, any non-defined field should be skipped by the parser so this seems ok.

Reply:
> > `optional NoiseExtensions extensions = 4;` is represented as a oneof in proto3
> 
> go-libp2p's older noise `.proto` file [with protobuf v3](https://github.com/libp2p/go-libp2p/blob/a82be49036096c93f44287780a8d32142b591976/p2p/security/noise/pb/payload.proto) doesn't have the `oneof` keyword. I don't think it's implicit if not set so I'm not sure this is correct. Am I missing something?

Sorry I didn't explain this fully. I'll explain more in a bit 


> > but it's a bit of a hack.
> 
> Is this because the field identifier is non-sequential? I think this is a feature of protocol buffers, any non-defined field should be skipped by the parser so this seems ok.

It's a hack because the real definition is proto2 but we're using a proto3 definition to workaround a protobuf-compiler bug.

Reply:
> In proto3 fields are optional by default so you can omit optional keywords.

@mpetrunic They're `singluar` by default so we need the `optional` keyword - https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules

Reply:
These two protobufs generate the same exact code (which is probably a separate bug since a proto2 encoder _should_ write empty bytes if set. See https://github.com/libp2p/specs/issues/465 for a discussion around proto2 vs proto3):

```
syntax = "proto2";

message NoiseExtensions {
    repeated bytes webtransport_certhashes = 1;
}

message NoiseHandshakePayload {
  optional bytes identity_key = 1;
  optional bytes identity_sig = 2;
  optional NoiseExtensions extensions = 4;
}
```

```
syntax = "proto3";

message NoiseExtensions {
    repeated bytes webtransport_certhashes = 1;
}

message NoiseHandshakePayload {
  bytes identity_key = 1;
  bytes identity_sig = 2;
  NoiseExtensions extensions = 4;
}
```

---

I think it may make sense to use proto3 long term here, but we still need to discuss that. For the mean time, I can set this as a proto3 definition that I've manually validated is compatible with proto2 parsers: https://github.com/MarcoPolo/proto2and3-playground/blob/9eb3987c04b348d64d825104813ad27c4df691e7/src/main.rs#L14